book/animations.md

```diff
@@ -5,56 +5,153 @@ prev: scheduling
+Browsers also use animations in response to user actions like
+scrolling, resizing, and pinch-zoom. And some types of animated edia,
+like videos, can also be included in web pages.[^video-anim] In this
+chapter we'll focus mostly on web page animations, though we'll touch
+on scrolling at the end.[^excuse]
+
+[^general-movement]: Here *movement* should be construed broadly to
+encompass all of the kinds of visual changes humans are used to seeing
+and good at recognizing---not just movement from side to side, but
+growing, shrinking, rotating, fading, blurring, and sharpening. The
+point is that an animation is not an *arbitrary* sequence of pictures;
+the sequence must feel, to a human mind trained by experience in the
+real world, to be a continuous motion.
 
 [animation]: https://en.wikipedia.org/wiki/Animation
 
-On web pages, there are several categories of common animations: DOM,
-input-driven and video-like. A DOM animation is a movement or visual effect
-change of elements on the screen, achieved by interpolating CSS properties of
-elements such as color, opacity or sizing, or changes of text content.
-Input-driven animations involve input of course: scrolling, page resizing,
-pinch-zoom, draggable menus and similar effects.[^video-anim]  In this chapter
-we'll focus mostly on DOM animations, with a bit of input-driven animations a
-bit at the end.[^excuse]
+[^video-anim]: Video-like animations also include animated images, and
+animated canvases. Since our browser doesn't support images, this
+topic is beyond the scope of this book, but it has its own
+[fascinating complexities][videong].
+
+[videong]: https://developer.chrome.com/blog/videong/
+
+Let's write a simple animation using the `requestAnimationFrame` API
+[implemented in Chapter 12](scheduling.md#animating-frames). This
+animation lets us request that some JavaScript run on the next
+*frame*, and we can have that JavaScript change the page slightly.
+To do this repeatedly, we'll need code like this:
+
+``` {.javascript file=example-opacity-js}
+function run_animation_frame() {
+    if (animate())
+        requestAnimationFrame(run_animation_frame);
+}
+requestAnimationFrame(run_animation_frame);
+```
 
-[^video-anim]: And video-like animations include videos, animated images, and
-animated canvases.
+Here `animate` makes some small change to the page to give the
+impression of continuous change. It returns `true` until it's done
+animating, and then stops. By changing what `animate` does we can
+change what animation occurs.
 
-[^excuse]: Video animations is a topic unto itself, but is beyond the
-scope of this book. Arguably, canvas is a bit different than the other two,
-since it's implemented by developer scripts. And of course a canvas can have
-animations within it that look about the same as some DOM animations.
+Let's write a fade animation. We can fade in something out by smoothly
+transitioning its `opacity` value from 0.0 to 0.999.[^why-not-one] If we
```

Comment:
The old version started opaque and faded out; this starts at transparent and fades in. Do you want it to start at 0.1? (I'm also not sure that 0.999 is necessary in this case, or if actually 0.001 opacity is important instead.)

Reply:
We'll I'd like the demo readers see to be a continuous animation, not a one-shot. And animating from one endpoint that is not completely invisible is better, which is why it was 0.1. 0.999 is necessary to avoid a composited layer pop on real browsers; 0.0 doesn't have the same problem.

Reply:
Good to know 0.0 doesn't have that problem. Why isn't there an optimization at that end? (To not draw things that are invisible?) I seem to recall that back in the day all the clickjacking attacks always used like 0.1% opacity for some reason, though perhaps that's optimizations in hit detection, not in compositing...

Reply:
There are optimizations for that 0.0. And you're right that there is no guarantee the composited layer won't be deleted.

Reply:
0.999 vs 1.0 causes a change of stacking context behavior though, which is another reason not to animate to 1.